payjoin-directory/src/db.rs

```diff
@@ -1,191 +1,105 @@
-    let (table_exists,): (bool,) =
-        sqlx::query_as("SELECT EXISTS (SELECT FROM pg_tables WHERE tablename = 'directory')")
-            .fetch_one(&pool)
-            .await?;
-
-    if !table_exists {
-        // Create the table
-        sqlx::query(
-            r#"
-            CREATE TABLE directory (
-                pubkey_id VARCHAR PRIMARY KEY,
-                req BYTEA,
-                res BYTEA
-            );
-        "#,
-        )
-        .execute(&pool)
-        .await?;
-
-        // Create the function for notification
-        sqlx::query(
-            r#"
-            CREATE OR REPLACE FUNCTION notify_change()
-            RETURNS TRIGGER AS $$
-            DECLARE
-                channel_name text;
-            BEGIN
-                channel_name := NEW.pubkey_id || '_' || TG_ARGV[0];
-                PERFORM pg_notify(channel_name, TG_TABLE_NAME || ', ' || NEW.pubkey_id);
-                RETURN NEW;
-            END;
-            $$ LANGUAGE plpgsql;
-        "#,
-        )
-        .execute(&pool)
-        .await?;
-
-        // Create triggers
-        sqlx::query(
-            r#"
-            CREATE TRIGGER directory_req_trigger
-            AFTER INSERT OR UPDATE OF req ON directory
-            FOR EACH ROW
-            EXECUTE FUNCTION notify_change('req');
-        "#,
-        )
-        .execute(&pool)
-        .await?;
-
-        sqlx::query(
-            r#"
-            CREATE TRIGGER directory_res_trigger
-            AFTER INSERT OR UPDATE OF res ON directory
-            FOR EACH ROW
-            EXECUTE FUNCTION notify_change('res');
-        "#,
-        )
-        .execute(&pool)
-        .await?;
+    pub async fn push_res(&self, pubkey_id: &str, data: Vec<u8>) -> RedisResult<()> {
+        self.push(pubkey_id, RES_COLUMN, data).await
     }
-    Ok(pool)
-}
 
-async fn push(
-    pool: &Pool<Postgres>,
-    pubkey_id: &str,
-    channel_type: &str,
-    data: Vec<u8>,
-) -> Result<(), sqlx::Error> {
-    // Use an UPSERT operation to insert or update the record
-    let query = format!(
-        "INSERT INTO directory (pubkey_id, {}) VALUES ($1, $2) \
-        ON CONFLICT (pubkey_id) DO UPDATE SET {} = EXCLUDED.{}",
-        channel_type, channel_type, channel_type
-    );
-
-    sqlx::query(&query).bind(pubkey_id).bind(data).execute(pool).await?;
-
-    Ok(())
-}
-
-async fn peek_with_timeout(
-    pool: &Pool<Postgres>,
-    pubkey_id: &str,
-    channel_type: &str,
-    timeout: Duration,
-) -> Option<Result<Vec<u8>, sqlx::Error>> {
-    tokio::time::timeout(timeout, peek(pool, pubkey_id, channel_type)).await.ok()
-}
+    pub async fn push(
+        &self,
+        pubkey_id: &str,
+        channel_type: &str,
+        data: Vec<u8>,
+    ) -> RedisResult<()> {
+        let mut conn = self.client.get_async_connection().await?;
+        let key = format!("{}:{}", pubkey_id, channel_type);
+        conn.set(&key, data.clone()).await?;
+        conn.publish(&key, data).await?;
```

Comment:
Nice detail oriented catch, this can push a unit value or a simple "updated" string instead.

Reply:
maybe its worth writing a test to demonstrate the different use cases?

Reply:
I think I am mainly confused about when the pub/sub will be utilized rather than the set/get. maybe iam missing something of how this should work in practice

Reply:
if the receiver is waiting for a payjoin psbt, it makes a GET request to the endpoint which calls peek. If there's no psbt yet then the redis::get will fail, but it'll subscribe for an update. If before the timeout ends the sender comes and posts a psbt which is redis::set, it'll also publish an event which the receiver is listening for. The receiver knows to redis::get again with a fresh value.